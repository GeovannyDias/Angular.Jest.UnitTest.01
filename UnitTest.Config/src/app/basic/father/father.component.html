<h1>Father Component</h1>
<button class="btn btn-primary" (click)="onSetClient('Geovanny')">Load Client</button>
<br><br>

<code class="mt-2">{{ client | json}}</code>

<hr>

<app-son [client]="client" (onDeleteClient)="client = undefined" (onClientUpdate)="client = $event"></app-son>