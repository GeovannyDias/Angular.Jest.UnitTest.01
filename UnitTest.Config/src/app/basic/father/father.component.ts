import { Component, OnInit } from '@angular/core';
import { ClientI } from '../interfaces/client.interface';

@Component({
  selector: 'app-father',
  templateUrl: './father.component.html',
  styleUrls: ['./father.component.scss']
})
export class FatherComponent implements OnInit {
  public client?: ClientI;
  constructor() { }

  ngOnInit(): void {
  }

  onSetClient(name: string) {
    this.client = { id: 1, name };
  }


}
