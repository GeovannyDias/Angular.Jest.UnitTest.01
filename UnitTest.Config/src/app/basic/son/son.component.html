<h5>Son Component</h5>
<code>{{ client | json }}</code>
<br><br>

<button *ngIf="client" class="btn btn-outline-primary m-2" (click)="onChangeId(5)" data-test="btn-change-id">Change ID</button>
<button *ngIf="client" class="btn btn-outline-danger m-2" (click)="onDelete()" data-test="btn-delete">Delete Client</button>