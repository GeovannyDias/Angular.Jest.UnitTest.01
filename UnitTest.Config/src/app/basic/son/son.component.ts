import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { ClientI } from '../interfaces/client.interface';

@Component({
  selector: 'app-son',
  templateUrl: './son.component.html',
  styleUrls: ['./son.component.scss']
})
export class SonComponent implements OnInit {
  @Input() client?: ClientI;
  @Output() onDeleteClient = new EventEmitter();
  @Output() onClientUpdate = new EventEmitter<ClientI>();

  constructor() { }

  ngOnInit(): void {
  }

  onChangeId(id: number) {
    // if (this.client) this.client.id = id; // No usar pasa por referencia y se actualiza en el fatherComponent
    if (this.client) {
      this.client = { ...this.client, id };
      this.onClientUpdate.emit({ ...this.client });
    }
  }

  onDelete() {
    this.client = undefined;
    this.onDeleteClient.emit();
  }

}
