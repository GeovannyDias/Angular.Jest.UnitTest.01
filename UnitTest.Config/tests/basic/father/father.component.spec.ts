import { ComponentFixture, TestBed } from '@angular/core/testing';
import { SonComponent } from '../../../src/app/basic/son/son.component';

import { FatherComponent } from '../../../src/app/basic/father/father.component';
import { By } from '@angular/platform-browser';

describe('FatherComponent', () => {
  let component: FatherComponent;
  let fixture: ComponentFixture<FatherComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [FatherComponent, SonComponent]
    })
      .compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FatherComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  // -----------------------------------------------------

  test('should do match to snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('should set client with name indicate', () => {
    component.onSetClient('Sofia-Test');
    fixture.detectChanges();

    const codeElm = compiled.querySelector('.mt-2');
    // console.log(codeElm?.textContent, typeof codeElm?.textContent);
    expect(codeElm?.textContent).toContain('"name"');
    expect(codeElm?.textContent).toContain('"Sofia-Test"');
  });

  test('should delete client, if emit onDeleteClient (Son)', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    const sonDebugElement = fixture.debugElement.query(By.directive(SonComponent)); // Toma el componente (Son) y se tiene el control del mismo
    const sonComponent: SonComponent = sonDebugElement.componentInstance; // 
    // console.log(sonComponent.client); // le asigna valor al padre y se la pasa al hijo
    sonComponent.onDeleteClient.emit();
    expect(component.client).toBe(undefined);
  });

  test('should update client with onClientUpdate', () => {
    component.client = { id: 1, name: 'Juan' };
    fixture.detectChanges();

    const sonDebugElement = fixture.debugElement.query(By.directive(SonComponent)); // Toma el componente (Son) y se tiene el control del mismo
    const sonComponent: SonComponent = sonDebugElement.componentInstance; // 

    sonComponent.onClientUpdate.emit({ id: 10, name: 'Juan' });
    expect(component.client).toEqual({ id: 10, name: 'Juan' });

  });


});
