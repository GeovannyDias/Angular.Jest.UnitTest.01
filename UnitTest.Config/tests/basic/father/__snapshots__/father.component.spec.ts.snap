// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FatherComponent should do match to snapshot 1`] = `
<div
  id="root1"
>
  <h1>
    Father Component
  </h1>
  <button
    class="btn btn-primary"
  >
    Load Client
  </button>
  <br />
  <br />
  <code
    class="mt-2"
  >
    
  </code>
  <hr />
  <app-son>
    <h5>
      Son Component
    </h5>
    <code>
      
    </code>
    <br />
    <br />
    
    
  </app-son>
</div>
`;
